### PR TITLE
picom-next: init at vNext

### PR DESCRIPTION
**Pull request by GKasparov**
###### Motivation for this change
Dual_Kawase blur isn't supported on the default picom

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

This PR adds the tryone fork of picom, which supports dual_kawase blur method.
Source: https://github.com/tryone144/picom/tree/feature/dual_kawase

### CODE REVIEW

**Comment by bobby285271** [comment]
Please add the package to `pkgs/top-level/all-packages.nix` and follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) when writing commit messages.

**Comment by GKasparov** [comment]
> An example on how to override the version directly from `pkgs/top-level/all-packages.nix` (can't link directly to GitHub link since this file is too big to GitHub to render it):
> 
> ```nix
>   libgit2_0_27 = libgit2.overrideAttrs (oldAttrs: rec {
>     version = "0.27.10";
>     src = fetchFromGitHub {
>       owner = "libgit2";
>       repo = "libgit2";
>       rev = "v${version}";
>       sha256 = "09jz2fzv0zl5058s0g1cpnw87a2rgg8wnjwlygi18i2n9nn6m0ad";
>     };
>   });
> ```

I've changed it to your suggestion, adding picom-next overrided with the next branch. Another way could be using the vNext tag rather than current committs, but I don't see the benefit in that

**Comment by thiagokokada** [comment]
> > An example on how to override the version directly from `pkgs/top-level/all-packages.nix` (can't link directly to GitHub link since this file is too big to GitHub to render it):
> > ```nix
> >   libgit2_0_27 = libgit2.overrideAttrs (oldAttrs: rec {
> >     version = "0.27.10";
> >     src = fetchFromGitHub {
> >       owner = "libgit2";
> >       repo = "libgit2";
> >       rev = "v${version}";
> >       sha256 = "09jz2fzv0zl5058s0g1cpnw87a2rgg8wnjwlygi18i2n9nn6m0ad";
> >     };
> >   });
> > ```
> 
> I've changed it to your suggestion, adding picom-next overrided with the next branch. Another way could be using the vNext tag rather than current committs, but I don't see the benefit in that

`vNext` tag AFAIK is always changing, so it is better to pin a commit like you did.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
 
   picom = callPackage ../applications/window-managers/picom {};
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+    version="vNext";
+      src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:28869]: This doesn't look like a valid commit.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
...
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+    maintainers = with maintainers; [ GKasparov ];
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:28872]:
```suggestion
    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
 
   picom = callPackage ../applications/window-managers/picom {};
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+    version="vNext";
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:28865]:
```suggestion
      version = "vNext";
```

Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:28865]: Not sure if we want it to have a proper version instead.

Comment (GKasparov) [reply pkgs/top-level/all-packages.nix:28865]: If so? which version
vNext indicates its the updated next branch. Idk either

Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:28865]: Not sure too, you can leave it as-is for now. Maybe some other reviewer has a better idea.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (GKasparov) [reply pkgs/top-level/all-packages.nix:28869]: Updated (I accidentally pasted the wrong string.

Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:28869]: Checks: https://github.com/yshui/picom/commit/fade045eadf171d2c732820d6ebde7d1943a1397

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -28861,6 +28861,17 @@ with pkgs;
...
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+    version = "vNext";
+      src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:28872]: Sorry :disappointed:.

```suggestion
    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```

Comment (GKasparov) [reply pkgs/top-level/all-packages.nix:28872]: Nws, my fault

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -28869,7 +28869,7 @@ with pkgs;
       rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
       sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
   };
-    maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
+    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
Comment (thiagokokada) [inline pkgs/top-level/all-packages.nix:28872]:
```suggestion
    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```

Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:28872]: This was where I said sorry because this was my mistake.

**Comment by SuperSandro2000** [comment]
I think this is more something for a local overlay and does not need to be included in nixpkgs.

**Comment by SuperSandro2000** [comment]
Also this should be in an extra file rather than in top-level.

**Comment by thiagokokada** [comment]
> I think this is more something for a local overlay and does not need to be included in nixpkgs.

Maybe. But I also think that `picom-next` has many features that some people wanted and it seems that the next stable release will still take sometime to happen. We can always add an alias `picom-next = picom` once the stable release happens.

See also: https://github.com/NixOS/nixpkgs/pull/142171#issuecomment-946837071

> Also this should be in an extra file rather than in top-level.

Any technical reason for this considering that there shouldn't be any differences in the derivation itself and also there are other examples of `overrideAtrrs` inside `all-packages.nix` :thinking: ?

**Comment by GKasparov** [comment]
> I think this is more something for a local overlay and does not need to be included in nixpkgs.

Like thiago, I know there are ways that users locally can install picom from the next branch but the dual_kawase blur feature and the patches in the next branch are really significant and the features and wanted, so why not automate the process for people seeking it by adding the package?

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -28861,6 +28861,18 @@ with pkgs;
 
   picom = callPackage ../applications/window-managers/picom {};
 
+  picom-next = picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+    version = "vNext";
```
Comment (jonringer) [inline pkgs/top-level/all-packages.nix:28866]:
```suggestion
    version = "unstable-YYYY-MM-DD";
```

in accordance with https://nixos.org/manual/nixpkgs/stable/#sec-package-naming

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:28872]: CC @GKasparov .

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (GKasparov) [reply pkgs/top-level/all-packages.nix:28866]: Resolved as indentation was fixed after moving the logic to it's own file

File: pkgs/applications/window-managers/picom/picom-next.nix [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+    version = "unstable-2021-11-01";
+    src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+      };
+    meta.maintainers = with maintainers; oldAttrs.maintainers ++ [ GKasparov ];
```
Comment (thiagokokada) [inline pkgs/applications/window-managers/picom/picom-next.nix:12]:
```suggestion
    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```

Comment (GKasparov) [reply pkgs/applications/window-managers/picom/picom-next.nix]: Committed.

File: pkgs/applications/window-managers/picom/picom-next.nix [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  pname = "picom-next";
+    version = "unstable-2021-11-01";
+    src = fetchFromGitHub {
+      owner = "yshui";
+      repo = "picom";
+      rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+      sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+      };
+    meta.maintainers = with maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
+  })
```
Comment (thiagokokada) [inline pkgs/applications/window-managers/picom/picom-next.nix:13]:
```suggestion
{ picom, lib, fetchFromGitHub }:

picom.overrideAttrs (oldAttrs: rec {
  pname = "picom-next";
  version = "unstable-2021-11-01";
  src = fetchFromGitHub {
    owner = "yshui";
    repo = "picom";
    rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
    sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
  };
  meta.maintainers = with lib.maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
})
```

Comment (GKasparov) [reply pkgs/applications/window-managers/picom/picom-next.nix]: Added lib import, changed to lib.maintainers and fixed indents

File: pkgs/applications/window-managers/picom/picom-next.nix [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+  version = "unstable-2021-11-01";
+  src = fetchFromGitHub {
+    owner = "yshui";
+    repo = "picom";
+    rev = "fade045eadf171d2c732820d6ebde7d1943a1397";
+    sha256 = "fPiLZ63+Bw5VCxVNqj9i5had2YLa+jFMMf85MYdqvHU=";
+  };
+  meta.maintainers = with lib.maintainers; oldAttrs.meta.maintainers ++ [ GKasparov ];
```
Comment (thiagokokada) [inline pkgs/applications/window-managers/picom/picom-next.nix:12]: You need to add yourself on `maintainers/maintainers-list.nix` file.

**Review by thiagokokada — COMMENTED** [review]
Code looks fine now.

Please add yourself to `maintainers/maintainers-list.nix` file (at alphabetical order) and squash your commits, and you should be good to go.

**Comment by thiagokokada** [comment]
> Code looks fine now.
> 
> Please add yourself to `maintainers/maintainers-list.nix` file (at alphabetical order) and squash your commits, and you should be good to go.

The final PR should have at most two commits:

- `maintainers: add GKasparov`
- `picom-next: init at unstable-2021-11-01`

File: pkgs/applications/window-managers/picom/picom-next.nix [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{ picom, lib, fetchFromGitHub }:
+
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-next";
+  version = "unstable-2021-11-01";
```
Comment (thiagokokada) [inline pkgs/applications/window-managers/picom/picom-next.nix:5]:
```suggestion
  version = "unstable-2021-10-31";
```

Date must be the date of fetched commit (not the date of the PR).

Comment (GKasparov) [reply pkgs/applications/window-managers/picom/picom-next.nix:5]: Sorry im in a different timezone so i got the date wrong

**Review by thiagokokada — COMMENTED** [review]
Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>picom-next</li>
  </ul>
</details>

Works now, please squash the commits as instructed and you're good to go.

**Comment by GKasparov** [comment]
> Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built:
> 
>     * picom-next
> 
> 
> Works now, please squash the commits as instructed and you're good to go.

How can I squash without write access. The drop down is usually next to the merge button

**Comment by thiagokokada** [comment]
> > Result of `nixpkgs-review pr 142171` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> > 1 package built:
> > ```
> > * picom-next
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Works now, please squash the commits as instructed and you're good to go.
> 
> How can I squash without write access. The drop down is usually next to the merge button

You need to squash manually in your fork and do a force push. Something like:

```
$ git remote add upstream https://github.com/NixOS/nixpkgs/
$ git fetch upstream
$ git rebase -i upstream/master # this will open your text editor so you can manually rebase your changes
$ git push --force
```

**Comment by GKasparov** [comment]
Hey, I've remade  the PR cleaner and with the proper commit messages. The way my branches were setup was extremely messy so please close this and merge the new one: https://github.com/NixOS/nixpkgs/pull/144696

**Pull request closed without merging**